design/src/rfcs/rfc0017_logging_sensitive.md

```diff
@@ -0,0 +1,403 @@
+
+These two cases illustrate that `smithy-rs` can only prevent violation of the specification in a restricted scope - logs emitted from generated code and the runtime crates. A `smithy-rs` specific guideline should be available to the customer which outlines how to avoid violating the specification in areas outside of our control.
+
+### Routing
+
+The sensitivity and HTTP bindings are declared within specific structures/operations. For this reason, in the general case, it's unknowable whether or not any given part of a request is sensitive until we determine which operation is tasked with handling the request and hence which fields are bound. Implementation wise, this means that any middleware applied _before_ routing has taken place cannot log anything sensitive without performing routing logic itself.
+
+Note that:
+
+- We are not required to deserialize the entire request before we can make judgments on what data is sensitive or not - only which operation it has been routed to.
+- We are permitted to emit logs prior to routing when:
+  - they contain no potentially sensitive data, or
+  - the request failed to route, in which case it's subject to the constraints of an operation.
+
+### Runtime Crates
+
+The crates existing in `rust-runtime` are not code generated - their source code is agnostic to the specific model in use. For this reason, if such a crate wanted to log potentially sensitive data then there must be a way to conditionally toggle that log without manipulation of the source code. Any proposed solution must acknowledge this concern.
+
+## Proposal
+
+This proposal serves to honor the sensitivity specification via code generation of a logging middleware which is aware of the sensitivity, together with a developer contract disallowing logging potentially sensitive data in the runtime crates. An internal and external guideline should be provided in addition to the middleware.
+
+All data known to be sensitive should be replaced with `"{redacted}"` when logged. Implementation wise this means that [tracing::Event](https://docs.rs/tracing/latest/tracing/struct.Event.html)s and [tracing::Span](https://docs.rs/tracing/latest/tracing/struct.Span.html)s of the form `debug!(field = "sensitive data")` and `span!(..., field = "sensitive data")` must become `debug!(field = "{redacted}")` and `span!(..., field = "{redacted}")`.
+
+### Debug Logging
+
+Developers might want to observe sensitive data for debugging purposes. It should be possible to opt-out of the redactions by enabling a feature flag `unredacted-logging` (which is disabled by default).
+
+To prevent excessive branches such as
+
+```rust
+if cfg!(feature = "unredacted-logging") {
+    debug!(%data, "logging here");
+} else {
+    debug!(data = "{redacted}", "logging here");
+}
+```
+
+the following wrapper should be provided
+
+```rust
+pub struct Sensitive<T>(T);
```

Comment:
Where should this struct live?

Reply:
Good question, there's conversation to be had here. I've raised it https://github.com/awslabs/smithy-rs/pull/1550#discussion_r923658147, but for the purpose of the RFC, perhaps it's specific enough to say that it'll live within the runtime crates.

Reply:
https://github.com/awslabs/smithy-rs/pull/1536/commits/55490389a1f66e3efad0b4db8f6a6babbc08fde9